tests/integration/test_charm.py

```diff
@@ -68,39 +69,54 @@ async def test_relations(self, ops_test: OpsTest):
-    # @pytest.mark.abort_on_fail
-    # @pytest.mark.timeout(1200)
-    # async def test_upgrade(self, ops_test: OpsTest):
-    #     await ops_test.model.remove_application(APP_NAME, block_until_done=True)
-    #
-    #     await ops_test.model.deploy(
-    #         APP_NAME, channel="ckf-1.7/stable", trust=True, config=OIDC_CONFIG
-    #     )
-    #     await ops_test.model.add_relation(f"{ISTIO_PILOT}:ingress", f"{APP_NAME}:ingress")
-    #     await ops_test.model.add_relation(
-    #         f"{ISTIO_PILOT}:ingress-auth", f"{APP_NAME}:ingress-auth"
-    #     )
-    #     await ops_test.model.add_relation(f"{APP_NAME}:oidc-client", f"{DEX_AUTH}:oidc-client")
-    #     await ops_test.model.applications[APP_NAME].set_config({"public-url": PUBLIC_URL})
-    #     await ops_test.model.wait_for_idle(
-    #         [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-    #         status="active",
-    #         raise_on_blocked=False,
-    #         raise_on_error=True,
-    #         timeout=600,
-    #     )
-    #
-    #     cmd = (
-    #         f"juju refresh {APP_NAME} "
-    #         f'--path "{pytest.charm_under_test}" --resource oci-image="{image_path}"'
-    #     )
-    #     await ops_test.run(*shlex.split(cmd))
-    #
-    #     await ops_test.model.wait_for_idle(
-    #         [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-    #         status="active",
-    #         raise_on_blocked=True,
-    #         raise_on_error=True,
-    #         timeout=600,
-    #     )
+    @pytest.mark.abort_on_fail
+    @pytest.mark.timeout(1200)
+    async def test_upgrade(self, ops_test: OpsTest):
+        """Test that charm can be upgraded from podspec to sidecar.
+
+        For this test we use 1.7/stable channel as the source for podspec charm.
+
+        Note: juju has a bug due to which you have to first scale podspec charm to 0,
+        then refresh, then scale up newly deployed app.
+        See https://github.com/juju/juju/pull/15701 for more info.
+        """
+        await ops_test.model.remove_application(APP_NAME, block_until_done=True)
+        print("Built OIDC App removed, deploy from stable channel")
+
+        await ops_test.model.deploy(
+            APP_NAME, channel="ckf-1.7/stable", trust=True, config=OIDC_CONFIG
+        )
+        await ops_test.model.add_relation(f"{ISTIO_PILOT}:ingress", f"{APP_NAME}:ingress")
+        await ops_test.model.add_relation(
+            f"{ISTIO_PILOT}:ingress-auth", f"{APP_NAME}:ingress-auth"
+        )
+        await ops_test.model.add_relation(f"{APP_NAME}:oidc-client", f"{DEX_AUTH}:oidc-client")
+        await ops_test.model.applications[APP_NAME].set_config({"public-url": PUBLIC_URL})
+
+        print("Stable charm is deployed, add relations")
+        await ops_test.model.wait_for_idle(
+            [APP_NAME, ISTIO_PILOT, DEX_AUTH],
+            status="active",
+            raise_on_blocked=False,
+            raise_on_error=True,
+            timeout=600,
+        )
```

Comment:
This entire setup is repeated almost verbatim in `test_relations`.
Would it better if we had a helper for that?

Reply:
not really. In another function we set config on DEX instead of APP

Reply:
Tha's because of the fact that the following is part of the APP setup. 
```python
await ops_test.model.applications[APP_NAME].set_config({"public-url": PUBLIC_URL})
```
You can move it to the top, right after deploying the stable charm, and then have a helper e.g. `configure_relations` to execute [the rest](https://github.com/canonical/oidc-gatekeeper-operator/blob/f73b4fcfcc03d18f10d888dd0ae097c8d97397e9/tests/integration/test_charm.py#L55-L70).

Reply:
Ok, then we have only relations and waiting for idle?

And then if waiting errors, I would receive abtraceback from helper function

It is pytest best practice that all assertions are done in the test function

Reply:
although, I am always for deduplication but I really do not see a value in this particular case. Extracting 3 relations into separate method does not bring that benefit

Reply:
> And then if waiting errors, I would receive abtraceback from helper function

That's true!

> It is pytest best practice that all assertions are done in the test function

And I get that! But in this case there are no assertions. We are just setting up relations with external charms.
TBH, the real benefit as far as I'm concerned is readability; it's hard to notice a diff (in this case configuring the APP instead of DEX, for the configuration of which we silently rely on a previous test) between so many copy-pasted lines, let alone maintain it if something changes in the relations configured.

If you feel strongly about it though, I won't block on this!

Reply:
So I understand you feel strongly about this 😂

Reply:
Yes :D

I wish we have a reasonable abstraction, but three lines of code doesn't worth it